bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,206 @@
+from collections.abc import Iterable
+
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.start_stop_node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._sr_nodes = {}
+        self._sr_cache = {}
+
+    def start(self, doc):
+        self.start_stop_node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        doc = dict(doc)
+        doc.pop("start", None)
+        metadata = dict(self.start_stop_node.metadata) | {"stop": doc}
+        self.start_stop_node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        desc_name = doc["name"]
+        metadata = dict(doc)
+
+        # Extract the variable fileds of the descriptor that can change during the run
+        uid = metadata.pop("uid")
+        conf_dict = {uid: metadata.pop("configuration", {})}
+        time_dict = {uid: metadata.pop("time", None)}
+        var_fields = {"configuration": conf_dict, "time": time_dict}
+
+        # Get or create a descriptor container
+        if desc_name in self.start_stop_node.keys():
+            desc_node = self.start_stop_nodes[desc_name]
+        else:
+            desc_node = self.start_stop_node.create_container(
+                key=desc_name, metadata=metadata
+            )
+            desc_node.create_container(key="external")
+            desc_node.create_container(key="internal")
+
+        # Update the metadata with the variable fields
```

Comment:
I like the way that this separates the variable and fixed information. That aspect is definitely an improvement.

I am not sure about the way this seems to be overwriting the variable information from Descriptor 1 with the variable information from Descriptor 2. If there are N events with one exposure time and then M events with a different exposure time, we need to write down both exposure times.

Reply:
Yes, I agree. The idea was to keep Configurations and Times from both descriptors and "overwrite" only the fixed information (it's going to be the same, so it doesn't matter -- there will be no overwriting effectively). That's why I define `conf_dict` and `time_dict` that store the variable information indexed with keys that are `uids` of specific descriptors. `pydantic.deep_update` will update these internal dictionaries and keep the previous values that have different keys.
Sorry, it probably wasn't clear because I reused the `metadata` variable and it looked like it was overwritten. I'll refactor this bit and add some comments.